pkg/southbound/gnmi/conn_manager.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2021-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gnmi
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/grpc"
+
+	"google.golang.org/grpc/connectivity"
+
+	topoapi "github.com/onosproject/onos-api/go/onos/topo"
+
+	"github.com/onosproject/onos-lib-go/pkg/errors"
+)
+
+// ConnManager gNMI connection manager
+type ConnManager interface {
+	Get(ctx context.Context, id ConnID) (Conn, error)
+	List(ctx context.Context) ([]Conn, error)
+	Watch(ctx context.Context, ch chan<- Conn) error
+	Connect(ctx context.Context, target *topoapi.Object) (Conn, error)
+	remove(connID ConnID) error
+}
+
+// NewConnManager creates a new gNMI connection manager
+func NewConnManager() ConnManager {
+	mgr := &connManager{
+		conns:       make(map[ConnID]Conn),
+		clientConns: make(map[topoapi.ID]*grpc.ClientConn),
+		eventCh:     make(chan Conn),
+	}
+	go mgr.processEvents()
+	return mgr
+}
+
+type connManager struct {
+	conns       map[ConnID]Conn
+	clientConns map[topoapi.ID]*grpc.ClientConn
+	connsMu     sync.RWMutex
+	watchers    []chan<- Conn
+	watchersMu  sync.RWMutex
+	eventCh     chan Conn
+}
+
+// Connect connecting to a gNMI target and adding a new gNMI connection
+func (m *connManager) Connect(ctx context.Context, target *topoapi.Object) (Conn, error) {
+	log.Infof("Connecting to the gNMI target: %s", target.ID)
+	conn, err := connect(target)
```

Comment:
We should only `connect` (and create a new `grpc.ClientConn`) if we know there’s not one already. This is creating a new client whether one existed already or not.